### ISSUE TITLE
Need to implement and test `emptyDir` performance on OSO 

### ISSUE THREAD

**Issue by ibuziuk** [redhat-developer/rh-che#883]
Good test case would be cloning big examples like - https://github.com/redhat-developer/rh-che/issues/772

**Comment by ibuziuk** [comment]
I tested quick-dirty implementation of `emptyDir` approach on minishift. The workspace pod looks the following way:

```
 Mounts:
      /projects from claim-che-workspace-8tuml36j (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-f28xq (ro)
      /workspace_logs from claim-che-workspace-erqvkccg (rw)
Conditions:
  Type           Status
  Initialized    True 
  Ready          True 
  PodScheduled   True 
Volumes:
  claim-che-workspace-8tuml36j:
    Type:    EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:  
  claim-che-workspace-erqvkccg:
    Type:    EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:  
  default-token-f28xq:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  default-token-f28xq
    Optional:    false
```

Time required for cloning the che to `/projects` (emptyDir):

![image](https://user-images.githubusercontent.com/1461122/45480733-f1ade080-b749-11e8-8721-2bd5d1b235d8.png)

Time required for cloning che to `~/git`(container file system)
![image](https://user-images.githubusercontent.com/1461122/45480919-71d44600-b74a-11e8-9c32-33fc12dedcb2.png)

**Comment by garagatyi** [comment]
Interesting, so container FS performance is better. hmm... I didn't expect that

**Comment by garagatyi** [comment]
@l0rd FYI

**Comment by l0rd** [comment]
It looks like writing on disk takes more on the volume filesystem but building with mvn takes more on the container filesystem. But yes I would have expected much better performances on the emptydir volume filesystem. It would be interesting understanding the storage driver that is used (device mapper?)

**Comment by ibuziuk** [comment]
@l0rd  when I was building che-starter a few times the average time of the builds in container FS vs `emptyDir` volume were pretty much the same. in terms of storage I believe by default `overlay2` is used in minishift https://github.com/minishift/minishift-centos-iso/blob/master/centos-7.template#L134

**Comment by l0rd** [comment]
Oh I thought you were testing on osio. In this case that's make sense: your volume is inside a VM shared folder that is much slower.

Anyway to get the storage driver: ‘docker info‘

**Comment by ibuziuk** [comment]
@l0rd yeah, that was tested on minishift, I would do the same tests on OSO once dev cluster would be setup properly

**Comment by ibuziuk** [comment]
@l0rd results from OSO 2a (https://console.starter-us-east-2a.openshift.com/) for the same set of tests:

```
    Mounts:
      /projects from claim-che-workspace-5ht6hjhv (rw)
      /tmp/workspace_logs from claim-che-workspace-93vv3wzm (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-kk4qf (ro)
Conditions:
  Type           Status
  Initialized    True 
  Ready          True 
  PodScheduled   True 
Volumes:
  claim-che-workspace-5ht6hjhv:
    Type:    EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:  
  claim-che-workspace-93vv3wzm:
    Type:    EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:  
```

Time required for cloning the che to /projects (emptyDir) ~ 10.2 sec :

![image](https://user-images.githubusercontent.com/1461122/45546753-8255f080-b81e-11e8-9096-82907ae75cd1.png)

Time required for cloning che to ~/git(container file system) ~ 10.9 secs
![image](https://user-images.githubusercontent.com/1461122/45546831-d19c2100-b81e-11e8-82bb-c0bd655c5950.png)

Building che-starter from /projects (emptyDir) with initially **empty** .m2 ~ 35.8s

![image](https://user-images.githubusercontent.com/1461122/45547443-2345ab00-b821-11e8-9f50-054d7996d3d9.png)

Building che-starter from ~/git(container file system) with **empty** .m2 ~ 39.2s

![image](https://user-images.githubusercontent.com/1461122/45547592-951df480-b821-11e8-8895-cdeeed3a635f.png)


Building che-starter from /projects (emptyDir) with initially **full** .m2 ~ 7.7s

![image](https://user-images.githubusercontent.com/1461122/45547502-44a69700-b821-11e8-942b-86d9a7f8509e.png)


Building che-starter from ~/git(container file system) with **full** .m2 ~ 7.8s

![image](https://user-images.githubusercontent.com/1461122/45547629-b383f000-b821-11e8-89f7-335c9ed5c0ea.png)

TL;DR: it appeared to be that there is no performance difference between `emptyDir` volume and container FS on OSO for `git clone` / `mvn clean install` operations. In general  all happens really fast in  comparison with local setup

**Comment by l0rd** [comment]
I can see that `emptyDir` is slightly faster and more important it allows sharing data (source code) among containers (which is critical when running tooling services in sidecars aka workspace.next)

**Comment by ibuziuk** [comment]
@l0rd wdyt, about having the first iteration of the per-workspace ephemeral support to be implemented via not mounting volume at all and using container FS + promoting this on openshift.io ASAP  and than enhance it with `emptyDir` support which would require more time and would be trickier to implement as per-workspace setting ? Those changes are planned to be done on rh-che side, so that we would not be dependent on the

**Comment by ibuziuk** [comment]
WIP pr with emptyDir volumes support

**Comment by ibuziuk** [comment]
@l0rd I quickly tried `EmptyDir` with `Memory` medium for git clone against minishift:
```
    Type:    EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:  Memory
```
building che-starter with empty `m2` took 01:31 min (without setting medium it took 8 minutes). Will try to do more tests tomorrow

**Comment by l0rd** [comment]
@ibuziuk that's promising! Anyway sorry for not answering before but yes I think implementing ephemeral workspaces using container FS is a good first step. 

Moving to emptyDir can be done as a second step because it is needed for workspace.next.

**Comment by garagatyi** [comment]
@ibuziuk can you elaborate what `Medium:  Memory` is?

**Comment by ibuziuk** [comment]
@garagatyi emptyDir.medium allows to mount tmpfs (RAM-backed filesystem) While tmpfs is very fast, files that are written will count against your Container’s memory limit.

**Comment by ibuziuk** [comment]
Closing this issue since the tests related to `emptyDir` volume on OSO have been done [1][2]

[1] https://github.com/redhat-developer/rh-che/issues/883#issuecomment-421330611
[2] https://github.com/redhat-developer/rh-che/issues/883#issuecomment-421428400

@l0rd should we create a separate issue for analyzing and supporting `Medium: Memory` option upstream ?

**Comment by garagatyi** [comment]
We should not use TMPFS I believe - we don't have that much memory for that in container.

**Comment by l0rd** [comment]
@ibuziuk @garagatyi yes memory is a concern. At least for OSIO. No need to open an issue now. Maybe in the future we will work on some PoC about having part of a workspace storage in memory for better perfs but that's not urgent.
